src/Compilers/CSharp/Portable/Binder/Binder_Crefs.cs

```diff
@@ -838,7 +836,12 @@ private Symbol ConstructWithCrefTypeParameters(int arity, TypeArgumentListSyntax
                 SeparatedSyntaxList<TypeSyntax> typeArgumentSyntaxes = typeArgumentListSyntax.Arguments;
                 var typeArgumentsWithAnnotations = ArrayBuilder<TypeWithAnnotations>.GetInstance(arity);
 
-                DiagnosticBag unusedDiagnostics = DiagnosticBag.GetInstance();
+                var unusedDiagnostics =
+#if DEBUG
+                    new BindingDiagnosticBag(DiagnosticBag.GetInstance());
```

Comment:
Why bother with this in debug mode?

Reply:
> Why bother with this in debug mode?

We assert in debug mode.

---
In reply to: [361551379](https://github.com/dotnet/roslyn/pull/40560#discussion_r361551379) [](ancestors = 361551379)

Reply:
> We assert in debug mode.

Then perhaps we ought to make `BindingDiagnosticBad.Discarded` always accumulate and assert in `Debug` mode? For the code I've gone through thus far this is the only place where we're specifically asserting stuff in debug mode, and I'm not sure what makes this location special vs all the rest of the places we're using that API.

---
In reply to: [361552560](https://github.com/dotnet/roslyn/pull/40560#discussion_r361552560) [](ancestors = 361552560,361551379)